### ISSUE TITLE
Query creation at destination failed: 403 Client Error: FORBIDDEN with `redash-migrate queries`

### ISSUE THREAD

**Issue by yu-iskw** [getredash/redash-toolbelt#106]
## Issue
I can't migrate queries with `redash-migrate queries` from the hosted redash to OSS redash v10.0.0, even though I was able to migrate data sources, groups, users and destinations.  When I tested migrating the hosted redash to a local docker-compose of the redash repository, `redash-migrate queries` worked. However, it doesn't work on the destination instance in Google Kubernetes Engine.

## Environment
- Python: 3.8.12
- redash: v10.0.0
- redash-toolbelt: both of 0.1.7 and 0.1.6

## Error
```bash
$ redash-migrate queries
Import queries...
Query 172644 - OK  - importing
Query 172644 - FAIL - Query creation at destination failed: 403 Client Error: FORBIDDEN for url: https://redash-qa.xxx.dev/api/queries
Query 173527 - OK  - importing
Query 173527 - FAIL - Query creation at destination failed: 403 Client Error: FORBIDDEN for url: https://redash-qa.xxx.dev/api/queries
...
```

##

**Comment by BalawalAli** [comment]
Hi, I am also encountering the same problem. Above 300 queries have been failed and do not migrate.

**Comment by yu-iskw** [comment]
I tried to forcefully activate users after executing `redash-migrate`, as I thought the issues is caused by inactivations of users. So, I updated the statuses of migrated uses in destination instance. However, that doesn't solve the issue. I am thinking of other possible solutions.

```sql
UPDATE
  users
SET
  details = (details::jsonb #- '{is_invitation_pending}')
WHERE
  (details->>'is_invitation_pending')::BOOL = true;
```

**Comment by yu-iskw** [comment]
Finally, I have solved the issue. We have to grant the `admin` permissins to users, because the redash client impersonates each users when creating queries. So, they must have the permission to create queries.

```sql
-- Group ID means the `admin` group.
UPDATE
  users
SET
  groups = groups || 1
WHERE
  NOT (1 = ANY(groups));
```

**Comment by susodapop** [comment]
> We have to grant the `admin` permissins to users,

Are you certain that `admin` was required? Did you try it with `default` first? 

Did you by chance not run `redash-migrate groups` before trying to import queries? It should not be necessary to manually manipulate the destination instance at all. The script handles all of this.

**Comment by yu-iskw** [comment]
In reality, the migrated users belong to the `default` group. However, creating queries with user's tokens didn't work.

**Comment by BalawalAli** [comment]
HI, We have successfully migrated to self-hosted redash. The above issue (in my scenario ) was due to DNS rules which were causing some queries status as Failed.

**Comment by yu-iskw** [comment]
@BalawalAli Interesting. Can you elaborate on that? I want to arrange my mind why migrating queries works after granting the `admin` group.

**Comment by ahonore42** [comment]
This fix worked for our instance after getting a similar 403 error. In this case it was for a specific admin user.

Error Example:
```
403 Client Error: FORBIDDEN for url: https://redash-qa.xxx.dev/api/users/2
```

1. Take the URL and use it to send a GET request in browser or API tool like Postman/Insomnia.

2. Locate the `api_key` in the return for your FORBIDDEN user. In our case, the forbidden user was an admin, so if they exist, make sure to give them admin permissions in the GUI if they don't have them.

Example return for the URL:

```js
{
  auth_type: "external",
  is_disabled: false,
  updated_at: "2021-11-04T16:52:34.270Z",
  profile_image_url: "https://somespicyprofile.net",
  is_invitation_pending: false,
  groups: [
    1,
    2
  ],
  id: 2,
  name: "Some Admin",
  created_at: "2021-11-01T21:27:00.882Z",
  disabled_at: null,
  is_email_verified: true,
  active_at: "2021-11-03T09:43:04Z",
  api_key: "AAAAAPPPPPPPPIIIIIkkkkkkkeeeeeeyyyyyyyy",
  email: "coolguy@somecompanyemail.com"
}
```

3. After locating the `api_key`, set it as the "destination_admin_api_key" in your `meta.json` file. Doing this fixed our 403 permissions issue.

**Comment by susodapop** [comment]
A 403 can happen for a  few reasons, but what I've found from running dozens of migrations in the last month: this typically happens because of the following situation:

1. At some point in the past, User A on the ORIGIN created query 1 against a data source 1.
2. Some time later, User A was removed from a group with permission to create queries against data source 1.
3. During migration, User A is copied to the DESTINATION with the `users` command. When the `groups` command is run, the user's group membership is synced from the ORIGIN.
4. During the migration, `redash-migrate` tries to copy query 1 with User A's API key. This is not allowed because User A does not have permission to create queries against data source 1.

The fix is to run `queries` before `groups`. This explains why @yu-iskw's solution worked.

This isn't the only case where a 403 could occur, but it's the most frequent case I've seen.
